src/layout/layout.rs

```diff
@@ -657,10 +677,16 @@ impl Layout {
         let flex = self.flex;
         let spacing = self.spacing;
         let constraints = &self.constraints;
+        let overlap = self.overlap;
 
         let area_size = Element::from((*variables.first().unwrap(), *variables.last().unwrap()));
         configure_area(&mut solver, area_size, area_start, area_end)?;
-        configure_variable_constraints(&mut solver, &variables, area_size)?;
+        configure_variable_in_area_constraints(&mut solver, &variables, area_size)?;
+        if overlap == 0 {
+            configure_variable_ascending_order_constraints(&mut solver, &variables)?;
+        } else {
+            configure_variable_overlap_constraints(&mut solver, &segments, overlap)?;
+        }
```

Comment:
Would be nice if this didn't have to be a special case. I'm surprised that it does have to be. What's different about having an overlap of 1 and an overlap of 0?

Reply:
overlap 1 only adds constraints for segments but overlap 0 adds constraints for all variables (i.e. segments and spacers).

Reply:
I think what you're saying here is that the overlap eats up any space that would have been allocated to the spacer, but the spacer doesn't eat any space allocated to the segments. Does that sound correct?

Reply:
To make sure we are on the same page, here's the terminology.

We have the following area:

```plain
<───────────────────────────────────area_width─────────────────────────────────>
┌─area_start                                                          area_end─┐
v                                                                              v
┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐
     │     Max(20)      │     │      Max(20)     │     │      Max(20)     │
└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘
```

We create elements `Vec<Element>` where each `Element` has a `start` and `end`:

```plain
┌────┬───────────────────┬────┬───element.start───┬────┬───────────────────┐
│    │                   │    │                   │    │                   │
v    v                   v    v                   v    v                   v
┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐
     │     Max(20)      │     │      Max(20)     │     │      Max(20)     │
└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘
    ^                   ^    ^                   ^    ^                   ^    ^
    │                   │    │                   │    │                   │    │
    └───────────────────┴────┴───element.end─────┴────┴───────────────────┴────┘
```

"segments" are `Vec<Element>` only for the areas of interest:

```plain
     ┌────────────────────────┬───segment.start────────┐
     │                        │                        │
     v                        v                        v
┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐
     │     Max(20)      │     │      Max(20)     │     │      Max(20)     │
└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘
                        ^                        ^                        ^
                        │                        │                        │
                        └──────────segment.end───┴────────────────────────┘
```

"spacers" are `Vec<Element>` the remaining gaps:

```plain
┌────────────────────────┬────────spacer.start────┬────────────────────────┐
│                        │                        │                        │
v                        v                        v                        v
┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐
     │     Max(20)      │     │      Max(20)     │     │      Max(20)     │
└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘
    ^                        ^                        ^                        ^
    │                        │                        │                        │
    └────────────────────────┴────spacer.end──────────┴────────────────────────┘
```

Reply:
yup - this aligns with my understanding. But adding this image is good for making it easy to visualize conversations, so thanks. The only thing I'd add to the description is that the variable's value really effectively points to the left side of the cell. This subtle distinction helps make it a bit more clear why the end is 1 more than the cell index, and makes the floating point values less ambiguous.

Reply:
> I think what you're saying here is that the overlap eats up any space that would have been allocated to the spacer, but the spacer doesn't eat any space allocated to the segments. Does that sound correct?

Previously, we were always adding this constraint:

```rust
    // all variables are in ascending order
    for (&left, &right) in variables.iter().tuple_windows() {
        solver.add_constraint(left | LE(REQUIRED) | right)?;
    }
    Ok(())
```

If we use individual variables:

```plain
┌────┬───────────────────┬────┬─────variables─────┬────┬───────────────────┬────┐
│    │                   │    │                   │    │                   │    │
v    v                   v    v                   v    v                   v    v
┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐
     │     Max(20)      │     │      Max(20)     │     │      Max(20)     │
└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘
^    ^                   ^    ^                   ^    ^                   ^    ^
└v1  └v2                 └v3  └v4                 └v5  └v6                 └v7  └v8
```

this does:

```
v1 <= v2
v2 <= v3
v3 <= v4
v4 <= v5
v5 <= v6
v6 <= v7
v7 <= v8
```

But when we have overlaps, we cannot do this anymore.

Reply:
yes, but we can do generically:

```
v1 - v2 <= 0 // not sure about this one
v2 - v3 <= 0
v3 - v4 <= overlap
v4 - v5 <= 0
v5 - v6 <= overlap
v6 - v7 <= 0
v7 - v8 <= 0 // not sure about this one either
```

Reply:
I _think_ we have to force `overlap`, i.e.

```
v3 - v4 == overlap
```

instead of 

```
v3 - v4 <= overlap
```

Otherwise we may get overlaps that are not what the user wants.